Add hybrid info overlay

ConnorMolz opened the pull request:
# Pull Request Template

## Description
I created a new overlay for the hybrid system energy state. Currently its only working for the iRacing GTP class cars, but its should be possible to convert it also to other sims 

## Related Issue
---

## Type of Change
- [x] New feature

## Checklist
- [x] My code follows the project's coding style.
- [x] I have tested my changes thoroughly.
- [x] My changes introduce no new warnings.

## Additional Notes
In testing

ConnorMolz committed:
Creating Overlay for Hybridsystem info

ConnorMolz committed:
Implement Data getting

RiddleTime commented on Race Element.Data/Games/iRacing/IRacingDataProvider.cs:
```diff
@@ -312,6 +314,7 @@ private void OnTelemetryData()
             float steeringRadians = -_iRacingSDK.Data.GetFloat(steeringWheelAngleDatum) * 2;
             float steeringPercentage = Single.RadiansToDegrees(steeringRadians) / localCar.Inputs.MaxSteeringAngle;
             localCar.Inputs.Steering = Math.Clamp(steeringPercentage, -1, 1);
+            localCar.Electronics.PushToPassLevel = _iRacingSDK.Data.GetFloat(pushToPassLevelDatum) * 100; // 0-1 to 0-100%
```
` * 100f`

RiddleTime commented on Race Element.HUD.Common/Overlays/Driving/HybridInfo/HybridInfoConfiguration.cs:
```diff
@@ -0,0 +1,61 @@
+
+namespace RaceElement.HUD.Common.Overlays.Driving.HybridInfo;
+
+public class HybridInfoConfiguration: OverlayConfiguration
```
`public sealed class HybridInfoConfiguration: OverlayConfiguration`

RiddleTime commented on Race Element.HUD.Common/Overlays/Driving/HybridInfo/HybridInfoOverlay.cs:
```diff
@@ -0,0 +1,180 @@
+    
+    public sealed override void BeforeStart()
+    {
+        try
```
remove the try catch, `BeforeStart` already has got a try catch clause when it is called within the app itself.

ConnorMolz replied:
Ok, I got this one form the LapDeltaOverlay class this one need then maybe some refactoring

RiddleTime commented on Race Element.HUD.Common/Overlays/Driving/HybridInfo/HybridInfoOverlay.cs:
```diff
@@ -0,0 +1,180 @@
+    Game = Game.iRacing
+)]
+
+public class HybridInfoOverlay: CommonAbstractOverlay
```
`public sealed class HybridInfoOverlay: CommonAbstractOverlay`

RiddleTime commented on Race Element.HUD.Common/Overlays/Driving/HybridInfo/HybridInfoOverlay.cs:
```diff
@@ -0,0 +1,180 @@
+        _font?.Dispose();
+    }
+
+    public override void Render(Graphics g)
```
` public sealed override void Render(Graphics g)`

RiddleTime commented on Race Element.HUD.Common/Overlays/Driving/HybridInfo/HybridInfoOverlay.cs:
```diff
@@ -0,0 +1,180 @@
+    public override void Render(Graphics g)
+    {
+        _cachedBackground?.Draw(g, 0, 0, _config.Bar.Width, _config.Bar.Height);
+        float energyLevel = GetEnergyLevel();
```
```csharp
float energyLevel = SimDataProvider.LocalCar.Electronics.PushToPassLevel;
```
then remove the `GetEnergyLevel()` function

ConnorMolz replied:
I took this method because maybe some other sims like LMU could use some other values or need some other calculations. But I can remove it if you want

RiddleTime replied:
It's a performance regression as it introduce an extra function call which is overhead.

RiddleTime reviewed (commented):
Added a couple of remarks.

Additionally, can this HUD Scale? 

Test it out by adding a constructor to the `HybridInfoConfiguration`, it will reveal a new hud option that allows you to Scale a HUD.
```csharp
 public HybridInfoConfiguration() => GenericConfiguration.AllowRescale = true;
 ```

RiddleTime commented on Race Element.HUD.Common/Overlays/Driving/HybridInfo/HybridInfoOverlay.cs:
```diff
@@ -0,0 +1,180 @@
+            {
+                Rectangle rect = new(0, 0, (int)(_config.Bar.Width / 2 * Scale), (int)(_config.Bar.Height * Scale));
+                using GraphicsPath path = GraphicsExtensions.CreateRoundedRectangle(rect, cornerRadius, 0, 0, cornerRadius);
+                g.FillPath(new SolidBrush(Color.FromArgb(_config.Colors.ThresholdHighOpacity, _config.Colors.ThresholdHighColor)), path);
```
make sure you also declare the brush like this in every Cached bitmap, do this for all the objects that inherit `IDisposable`. Even `Pen` inherits this.
```csharp
Using SolidBrush brush = new(Color.FromArgb(_config.Colors.ThresholdHighOpacity, config.Colors.ThresholdHighColor);
g.FillPath(brush, path);
```

RiddleTime commented on Race Element.HUD.Common/Overlays/Driving/HybridInfo/HybridInfoOverlay.cs:
```diff
@@ -0,0 +1,180 @@
+    
+    
+    public HybridInfoOverlay(Rectangle rectangle) : base(rectangle, "Hybrid Info")
+    {
```
You can move everything in here to `BeforeStart`

ConnorMolz committed:
make multiplicator to a float

ConnorMolz committed:
Add the wanted changes (mostly set classes and methods sealed)

ConnorMolz committed:
Add Generic Config to the overlay

ConnorMolz committed:
refactor _cachedEnergyStage(n) Color including

ConnorMolz committed:
Move most of the consturctor to beforeStart

ConnorMolz commented:
So I updated the code so far

The constructor is now only including the _font because its readonly and need to be there.

I add the comment of the GetEnergyLevel() suggestion (https://github.com/RiddleTime/Race-Element/pull/284#discussion_r2600163059) 

I try to test it today. If not today then on Thursday.

ConnorMolz committed:
Implement suggestion and fix small Bug

RiddleTime commented on Race Element.HUD.Common/Overlays/Driving/HybridInfo/HybridInfoOverlay.cs:
```diff
@@ -0,0 +1,177 @@
+    
+    public sealed override void BeforeStop()
+    {
+        Width = _config.Bar.Width + 1;
```
I think you want to place this in `BeforeStart()`

ConnorMolz committed:
Fix small mistake

ConnorMolz commented:
Fix the mistake where I put the constructor content in the wrong method

I can send you a Screenshot on Thursday.

ConnorMolz commented:
Also can you remove my old Discord Account from the Discord Server?
I lost access mid year, I had a new one I named it the same on the Server.

ConnorMolz commented:
> > Fix the mistake where I put the constructor content in the wrong method
> > I can send you a Screenshot on Thursday.
> 
> No rush, think I've added you now on discord.

Yeah, thank you :)

actions-user committed:
Update contributors list (excluding GitHub-Action)

RiddleTime reviewed (changes requested):
- Please have a look at your code and look for unused variables or class properties, (see the Configuration class for an example)
- Test the HUD so it can Scale, users will complaign if it can't, in an earlier review I've left a tip on how to enable the scaling option

ConnorMolz committed:
Merge branch 'RiddleTime:dev' into dev

ConnorMolz committed:
Merge pull request #2 from ConnorMolz/dev

set Branch to updated version

ConnorMolz committed:
Fix scaling

ConnorMolz committed:
Cleanup imports and cleanup variables

RiddleTime committed:
hybrid info hud: add preview data

- added option to adjust energy level in preview image
- added option to adjust the refreshrate

RiddleTime committed:
refactor hybridinfo overlay

- fix corner radius (from config)
- add 'using pen' to improve garbage collection
- removed parameters for energyLevel and replaced it with private member _energyLevel

RiddleTime committed:
added another using

(almost missed this one 😅 )

RiddleTime committed:
Merge branch 'dev' into Add-hybrid-info-overlay

RiddleTime committed:
refactor

RiddleTime committed:
Merge branch 'Add-hybrid-info-overlay' of https://github.com/ConnorMolz/Race-Element into pr/284
